File: pixels-core/src/main/java/io/pixelsdb/pixels/core/writer/PixelsWriterOption.java
/*
 * Copyright 2023 PixelsDB.
 *
 * This file is part of Pixels.
 *
 * Pixels is free software: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Pixels is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Affero GNU General Public License for more details.
 *
 * You should have received a copy of the Affero GNU General Public
 * License along with Pixels.  If not, see
 * <https://www.gnu.org/licenses/>.
 */
package io.pixelsdb.pixels.core.writer;

import io.pixelsdb.pixels.core.encoding.EncodingLevel;

import java.nio.ByteOrder;

/**
 * @author hank
 * @create 2023-08-14
 */
public class PixelsWriterOption
{
    /**
     * The number of values in each pixel.
     */
    private int pixelStride;
    private EncodingLevel encodingLevel;
    private ByteOrder byteOrder;
    /**
     * Whether nulls positions in column are padded by arbitrary values and occupy storage and memory space.
     */
    private boolean nullsPadded;

    public PixelsWriterOption() { }

    public int getPixelStride()
    {
        return pixelStride;
    }

    public PixelsWriterOption pixelStride(int pixelStride)
    {
        this.pixelStride = pixelStride;
        return this;
    }

    public EncodingLevel getEncodingLevel()
    {
        return encodingLevel;
    }

    public PixelsWriterOption encodingLevel(EncodingLevel encodingLevel)
    {
        this.encodingLevel = encodingLevel;
        return this;
    }

    public ByteOrder getByteOrder()
    {
        return byteOrder;
    }

    public PixelsWriterOption byteOrder(ByteOrder byteOrder)
    {
        this.byteOrder = byteOrder;
        return this;
    }

    public boolean isNullsPadded()
    {
        return nullsPadded;
    }

    public PixelsWriterOption nullsPadded(boolean nullsPadded)
    {
        this.nullsPadded = nullsPadded;
        return this;
    }
}

File: cpp/pixels-core/lib/writer/PixelsWriterOption.cpp
/*
 * Copyright 2024 PixelsDB.
 *
 * This file is part of Pixels.
 *
 * Pixels is free software: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Pixels is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Affero GNU General Public License for more details.
 *
 * You should have received a copy of the Affero GNU General Public
 * License along with Pixels.  If not, see
 * <https://www.gnu.org/licenses/>.
 */

//
// Created by gengdy on 24-11-25.
//

#include "writer/PixelsWriterOption.h"
#include <iostream>

PixelsWriterOption::PixelsWriterOption() {}

int PixelsWriterOption::getPixelsStride() const {
    return this->pixelsStride;
}

std::shared_ptr<PixelsWriterOption> PixelsWriterOption::setPixelsStride(int pixelsStride) {
    this->pixelsStride = pixelsStride;
    return shared_from_this();
}

EncodingLevel PixelsWriterOption::getEncodingLevel() const {
    return this->encodingLevel;
}

std::shared_ptr<PixelsWriterOption> PixelsWriterOption::setEncodingLevel(EncodingLevel encodingLevel) {
    this->encodingLevel = encodingLevel;
    return shared_from_this();
}

bool PixelsWriterOption::isNullsPadding() const {
    return this->nullsPadding;
}

std::shared_ptr<PixelsWriterOption> PixelsWriterOption::setNullsPadding(bool nullsPadding) {
    this->nullsPadding = nullsPadding;
    return shared_from_this();
}

ByteOrder PixelsWriterOption::getByteOrder() const {
    return byteOrder;
}

void PixelsWriterOption::setByteOrder(ByteOrder byte_order) {
    byteOrder = byte_order;
}
